### PR TITLE
Add MemifConnect, Flush composites into /sdk/vppagent

### PR DESCRIPTION
**Pull request by onobrod**
## Description
Added vppagent package into sdk. Added MemifConnect and Flush composites.

MemifConnect creates a DataChange with Memif interface and stores it in opaque data.
Flush reads DataChange from opaque and writes it into vpp-agent.

## Motivation and Context
https://github.com/networkservicemesh/networkservicemesh/issues/1275 https://github.com/networkservicemesh/networkservicemesh/issues/1276

## How Has This Been Tested?

- [x] Covered by existing integration testing
- [ ] Added integration testing to cover
- [x] Tested locally
- [ ] Have not tested

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by onobrod** [comment]
Some review comments related to this PR are added to #1304 already. Not sure if it is really useful to have two dependent PRs, maybe I should close this PR and leave only one.

File: sdk/vppagent/flush.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+type Flush struct {
+	endpoint.BaseCompositeEndpoint
+	Conn     *grpc.ClientConn
+}
+
+// Request implements the request handler
+func (f *Flush) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*connection.Connection, error) {
+
+	if f.GetNext() == nil {
+		logrus.Fatal("The VPP Agent Flush composite requires that there is Next set")
```
Comment (haiodo) [inline sdk/vppagent/flush.go]: Should be better to return error with message?

File: sdk/vppagent/flush.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+	incomingConnection, err := f.GetNext().Request(ctx, request)
+	if err != nil {
+		logrus.Errorf("Next request failed: %v", err)
+		return nil, err
+	}
+
+	opaque := f.GetNext().GetOpaque(incomingConnection)
+	if opaque == nil {
+		err = fmt.Errorf("received empty data from Next")
```
Comment (haiodo) [inline sdk/vppagent/flush.go]: I suppose we just need to return message and not print it here.

File: sdk/vppagent/flush.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+// Request implements the request handler
+func (f *Flush) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*connection.Connection, error) {
+
+	if f.GetNext() == nil {
+		logrus.Fatal("The VPP Agent Flush composite requires that there is Next set")
+	}
+
+	incomingConnection, err := f.GetNext().Request(ctx, request)
+	if err != nil {
+		logrus.Errorf("Next request failed: %v", err)
```
Comment (haiodo) [inline sdk/vppagent/flush.go]: Should we just return error without printing it?

File: sdk/vppagent/flush.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	client := configurator.NewConfiguratorClient(f.Conn)
+
+	if _, err := client.Delete(ctx, &configurator.DeleteRequest{Delete: dataChange}); err != nil {
+		logrus.Error(err)
+		return err
+	}
+	return nil
+}
+
+func (f *Flush) reset() error {
```
Comment (haiodo) [inline sdk/vppagent/flush.go]: Why it is named reset?

File: sdk/vppagent/memif_connect.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	Workspace      string
+	ConnectionSide ConnectionSide
+	Connections    map[string]*ConnectionData
+}
+
+// Request implements the request handler
+func (mc *MemifConnect) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*connection.Connection, error) {
+
+	if mc.GetNext() == nil {
+		logrus.Fatal("The VPP Agent Memif Connect composite requires that there is Next set")
```
Comment (haiodo) [inline sdk/vppagent/memif_connect.go]: Should we return error here?

File: sdk/vppagent/flush.go (hunk 3 above)
Comment (edwarnicke) [reply sdk/vppagent/flush.go]: Agreed.

File: sdk/vppagent/flush.go (hunk 2 above)
Comment (edwarnicke) [reply sdk/vppagent/flush.go]: Agreed

File: sdk/vppagent/flush.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+		return nil, err
+	}
+
+	opaque := f.GetNext().GetOpaque(incomingConnection)
+	if opaque == nil {
+		err = fmt.Errorf("received empty data from Next")
+		logrus.Errorf("Unable to find the DataChange: %v", err)
+		return nil, err
+	}
+	dataChange := opaque.(*ConnectionData).DataChange
```
Comment (edwarnicke) [inline sdk/vppagent/flush.go]: You should have a proper check for this type conversion:

https://tour.golang.org/methods/15

also check for and return an error if DataChange is nil.

File: sdk/vppagent/flush.go [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+	}
+
+	return incomingConnection, nil
+}
+
+// Close implements the close handler
+func (f *Flush) Close(ctx context.Context, connection *connection.Connection) (*empty.Empty, error) {
+	opaque := f.GetNext().GetOpaque(connection)
+	if opaque != nil {
+		dataChange := opaque.(*ConnectionData).DataChange
```
Comment (edwarnicke) [inline sdk/vppagent/flush.go]: Check the type assertion and nil of DataChange and properly handle the errors if those checks fail.

File: sdk/vppagent/memif_connect.go (hunk 5 above)
Comment (edwarnicke) [reply sdk/vppagent/memif_connect.go]: Agreed.

**Comment by onobrod** [comment]
Thanks to GitHub it becomes almost impossible to maintain two dependent PRs. I cannot construct a proper link showing file diff — GitHub get crazy because I have new commits in both branches. Will close this PR when resolve all conversations.

File: sdk/vppagent/flush.go (hunk 4 above)
Comment (onobrod) [reply sdk/vppagent/flush.go]: 1) It was named like that before:
https://github.com/networkservicemesh/networkservicemesh/blob/master/test/applications/cmd/vppagent-icmp-responder-nse/vppagent.go#L56
https://github.com/networkservicemesh/networkservicemesh/blob/c3e247e9a672330247b13dcbfe3eb416e2689a9b/test/applications/cmd/vppagent-firewall-nse/vppagent.go#L77

2) It resets VPP Agent config

File: sdk/vppagent/flush.go (hunk 1 above)
Comment (onobrod) [reply sdk/vppagent/flush.go]: Fixed

File: sdk/vppagent/flush.go (hunk 3 above)
Comment (onobrod) [reply sdk/vppagent/flush.go]: Fixed

File: sdk/vppagent/flush.go (hunk 2 above)
Comment (onobrod) [reply sdk/vppagent/flush.go]: Fixed

File: sdk/vppagent/memif_connect.go (hunk 5 above)
Comment (onobrod) [reply sdk/vppagent/memif_connect.go]: Fixed

File: sdk/vppagent/flush.go (hunk 6 above)
Comment (onobrod) [reply sdk/vppagent/flush.go]: Fixed

File: sdk/vppagent/flush.go (hunk 7 above)
Comment (onobrod) [reply sdk/vppagent/flush.go]: Fixed

**Comment by onobrod** [comment]
Fixed all issues in this PR. Please, use #1304 for review, closing this.

**Pull request closed without merging**